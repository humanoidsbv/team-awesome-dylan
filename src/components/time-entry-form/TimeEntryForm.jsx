import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';

import IconClose from '../../assets/icons/icon-close.svg';
import styles from './TimeEntryForm.module.css';

function TimeEntryForm({ createTimeEntry }) {
  const today = new Date()
    .toISOString()
    .split('T')
    .shift();

  const [activity, setActivity] = useState('');
  const [client, setClient] = useState('');
  const [date, setDate] = useState(today);
  const [startTime, setStartTime] = useState('09:00');
  const [stopTime, setStopTime] = useState('17:00');
  const [formState, setFormState] = useState(true);
  const [validity, setValidity] = useState({
    clientVhStateElement: true,
    activityVhStateElement: true
  });

  const handleBlur = event => {
    console.log(event.target.name);
    console.log(event.target.checkValidity());
    setValidity({
      ...validity,
      [event.target.name]: event.target.checkValidity()
    });
  };

  const formRef = useRef(null);
  console.log(formRef);

  const handleSubmit = event => {
    event.preventDefault();

    const formStatus = formRef.current.checkValidity();
    console.log(formStatus);
    setFormState(formStatus);
    if (formStatus === false) return false;

    createTimeEntry({
      client,
      id: Math.random(),
      startTimestamp: new Date(`${date}T${startTime}`).toISOString(),
      stopTimestamp: new Date(`${date}T${stopTime}`).toISOString()
    });
  };

  return (
    <form
      className={`${styles.timeEntryForm} ${
        formState ? styles.timeEntryForm : styles.timeEntryFormInvalid
      }`}
      onSubmit={handleSubmit}
      ref={formRef}
    >
      <IconClose className={styles.iconClose} />
      <label className={styles.labelClient} htmlFor="client">
        CLIENT
        <input
          className={`${styles.inputClientValid} ${
            validity.clientVhStateElement
              ? styles.inputClientValid
              : styles.inputClientInvalid
          }`}
          id="client"
          maxLength="35"
          minLength="2"
          name="clientVhStateElement"
          onBlur={handleBlur}
          onChange={({ target }) => setClient(target.value)}
          placeholder="--enter a client--"
          required
          value={client}
        />
      </label>
      <label className={styles.labelActivity} htmlFor="activity">
        ACTIVITY
        <input
          className={`${styles.inputActivityValid} ${
            validity.activityVhStateElement
              ? styles.inputActivityValid
              : styles.inputActivityInvalid
          }`}
          id="activity"
          maxLength="35"
          minLength="2"
          name="activityVhStateElement"
          onBlur={handleBlur}
          onChange={({ target }) => setActivity(target.value)}
          placeholder="--enter an activity--"
          required
          value={activity}
        />
      </label>
      <label className={styles.labelDate} htmlFor="date">
        DATE
        <input
          className={styles.inputDate}
          id="date"
          name="DateVhStateElement"
          onBlur={handleBlur}
          onChange={({ target }) => setDate(target.value)}
          required
          type="date"
          value={date}
        />
      </label>
      <label className={styles.labelsTime} htmlFor="startTime">
        FROM
        <input
          className={styles.inputsTime}
          id="startTime"
          name="fromTimeVhStateElement"
          onBlur={handleBlur}
          onChange={({ target }) => setStartTime(target.value)}
          required
          type="time"
          value={startTime}
        />
      </label>
      <label className={styles.labelsTime} htmlFor="stopTime">
        TO
        <input
          className={styles.inputsTime}
          id="endTime"
          name="toTimeVhStateElement"
          onBlur={handleBlur}
          onChange={({ target }) => setStopTime(target.value)}
          required
          type="type"
          value={stopTime}
        />
      </label>
      <button
        className={`${styles.buttonTimeEntry} ${
          validity.activityVhStateElement
            ? styles.buttonTimeEntryValid
            : styles.buttonTimeEntryInvalid
        }`}
        type="submit"
      >
        Add
      </button>
    </form>
  );
}

TimeEntryForm.propTypes = {
  createTimeEntry: PropTypes.func.isRequired
};

export default TimeEntryForm;
