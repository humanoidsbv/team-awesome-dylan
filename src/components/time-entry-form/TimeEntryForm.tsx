import React, { useState, useRef } from 'react';

import IconClose from '../../assets/icons/icon-close.svg';
import styles from './TimeEntryForm.module.css';
import { TimeEntriesProps } from '../../redux/time-entries/types';
import { ValidityState } from '../../shared/types';

interface TimeEntryFormProps extends TimeEntriesProps {
  isTimeEntryFormVisible: boolean;
  toggleTimeEntryForm: () => void;
}

function TimeEntryForm({
  clients,
  createTimeEntry,
  isTimeEntryFormVisible,
  toggleTimeEntryForm
}: TimeEntryFormProps): React.ReactElement {
  const today = new Date()
    .toISOString()
    .split('T')
    .shift();

  const [activity, setActivity] = useState('');
  const [client, setClient] = useState('');
  const [date, setDate] = useState(today);
  const [startTime, setStartTime] = useState('09:00');
  const [stopTime, setStopTime] = useState('17:00');
  const [validity, setValidity] = useState<ValidityState>({});

  const formRef = useRef(null);

  const handleBlur = (event: React.ChangeEvent<HTMLSelectElement | HTMLInputElement>): void => {
    setValidity({
      ...validity,
      form: formRef.current.checkValidity(),
      [event.target.name]: event.target.checkValidity()
    });
  };

  const handleCancel = (event): void => {
    event.preventDefault();
    toggleTimeEntryForm();
  };

  const handleSubmit = (event): void => {
    event.preventDefault();

    setClient('');
    setActivity('');

    createTimeEntry({
      client: Number(client),
      id: Math.random(),
      startTimestamp: new Date(`${date}T${startTime}`).toISOString(),
      stopTimestamp: new Date(`${date}T${stopTime}`).toISOString()
    });
    toggleTimeEntryForm();
  };

  return (
    <form
      className={`${styles.timeEntryForm} ${!isTimeEntryFormVisible
        && styles.timeEntryFormHide}`}
      onSubmit={handleSubmit}
      ref={formRef}
    >
      <IconClose className={styles.iconClose} onClick={handleCancel} />
      <label className={styles.labelClient} htmlFor="client">
        CLIENT
        <select
          className={styles.inputClient}
          id="client"
          name="client"
          onBlur={handleBlur}
          onChange={({ target }): void => setClient(target.value)}
          required
          value={client}
        >
          <option className={styles.inputClient} disabled value="">
            --select a client--
          </option>
          {clients.map(({ clientName, id }) => (
            <option key={id} value={id}>
              {clientName}
            </option>
          ))}
        </select>
      </label>
      <label className={styles.labelActivity} htmlFor="activity">
        ACTIVITY
        <input
          className={`${styles.inputActivityValid} ${
            validity.activity === false
              ? styles.inputActivityInvalid
              : styles.inputActivityValid
          }`}
          id="activity"
          maxLength={35}
          minLength={2}
          name="activity"
          onBlur={handleBlur}
          onChange={({ target }): void => setActivity(target.value)}
          placeholder="--enter an activity--"
          required
          value={activity}
        />
      </label>
      <label className={styles.labelDate} htmlFor="date">
        DATE
        <input
          className={styles.inputDate}
          id="date"
          name="date"
          onBlur={handleBlur}
          onChange={({ target }): void => setDate(target.value)}
          required
          type="date"
          value={date}
        />
      </label>
      <label className={styles.labelsTime} htmlFor="startTime">
        FROM
        <input
          className={styles.inputsTime}
          id="startTime"
          name="startTime"
          onBlur={handleBlur}
          onChange={({ target }): void => setStartTime(target.value)}
          required
          type="time"
          value={startTime}
        />
      </label>
      <label className={styles.labelsTime} htmlFor="stopTime">
        TO
        <input
          className={styles.inputsTime}
          id="stopTime"
          name="stopTime"
          onBlur={handleBlur}
          onChange={({ target }): void => setStopTime(target.value)}
          required
          type="type"
          value={stopTime}
        />
      </label>
      <button
        className={`${styles.buttonTimeEntry}`}
        type="submit"
        disabled={validity.form !== true}
      >
        Add
      </button>
    </form>
  );
}

export default TimeEntryForm;
