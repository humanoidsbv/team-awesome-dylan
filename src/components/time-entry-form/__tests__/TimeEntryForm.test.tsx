import * as React from 'react';
import { shallow } from 'enzyme';

import TimeEntryForm from '../TimeEntryForm';

it('matches snapshot', () => {
  const timeEntryForm = shallow(
    <TimeEntryForm
      isTimeEntryFormVisible={true}
      toggleTimeEntryForm={() => {}}
      clients ={[{
        branch:'test',
        clientName:'test',
      }]}
      createTimeEntry={({}) => null}
    />
  );
  expect(timeEntryForm).toMatchSnapshot();
});
