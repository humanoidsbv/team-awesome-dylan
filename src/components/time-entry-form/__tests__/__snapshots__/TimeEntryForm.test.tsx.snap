// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form
  className="timeEntryForm false"
  onSubmit={[Function]}
>
  <test-file-stub
    className="iconClose"
    onClick={[Function]}
  />
  <label
    className="labelClient"
    htmlFor="client"
  >
    CLIENT
    <select
      className="inputClient"
      id="client"
      name="client"
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      value=""
    >
      <option
        className="inputClient"
        disabled={true}
        value=""
      >
        --select a client--
      </option>
      <option
        key="123"
        value={123}
      >
        test
      </option>
    </select>
  </label>
  <label
    className="labelActivity"
    htmlFor="activity"
  >
    ACTIVITY
    <input
      className="inputActivityValid inputActivityValid"
      id="activity"
      maxLength={35}
      minLength={2}
      name="activity"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="--enter an activity--"
      required={true}
      value=""
    />
  </label>
  <label
    className="labelDate"
    htmlFor="date"
  >
    DATE
    <input
      className="inputDate"
      id="date"
      name="date"
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      type="date"
      value="2019-10-17"
    />
  </label>
  <label
    className="labelsTime"
    htmlFor="startTime"
  >
    FROM
    <input
      className="inputsTime"
      id="startTime"
      name="startTime"
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      type="time"
      value="09:00"
    />
  </label>
  <label
    className="labelsTime"
    htmlFor="stopTime"
  >
    TO
    <input
      className="inputsTime"
      id="stopTime"
      name="stopTime"
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      type="type"
      value="17:00"
    />
  </label>
  <button
    className="buttonTimeEntry"
    disabled={true}
    type="submit"
  >
    Add
  </button>
</form>
`;
